File: merge/merge.pl
#!/usr/bin/perl
# @wxian2016Mar03

# == usage
die "Usage: perl $0 <file_prefix>" unless (@ARGV ==1);

my $time=localtime();
print ">>start at $time\n";

# == path
our $file_prefix = $ARGV[0];
our $GENE = "/lustre/project/og04/pub/database/onco_panel/gene";
our $ANNO = "/lustre/project/og04/shenzhongji/human_tumor/QuanlityAssessment/anno/";
#our $ANNO="/export/home/wangxian/test.pl/";
our $OUTPUT_GENE="/lustre/project/og04/wangxian/merge/".$file_prefix.".gene.snpeff.vcf";
our $UNION="/lustre/project/og04/wangxian/merge/".$file_prefix.".var.union";
#our $OUTPUT_GENE="/export/home/wangxian/test.pl/".$file_prefix.".gene.snpeff.vcf";
#our $UNION="/export/home/wangxian/test.pl/".$file_prefix.".union";
our $SNPEFF=$ANNO.$file_prefix.".var.snpeff.vcf";
our $INDEL_ANNODB=$ANNO.$file_prefix.".var.annodb.indel.exome_summary.csv";
our $SNP_ANNODB=$ANNO.$file_prefix.".var.annodb.snp.exome_summary.csv";
our $CLINVAR_ANNOVAR=$ANNO.$file_prefix.".var.annovar.hg19_clinvar_20150629_dropped";
our $COSMIC70_ANNOVAR=$ANNO.$file_prefix.".var.annovar.hg19_cosmic70_dropped";
our $ONCOTATOR=$ANNO.$file_prefix.".var.oncotator.vcf";

# == build index of gene
open (GENE, '<', $GENE);
our %hash_gene=();
while (<GENE>) {
	chomp;
	my $key=$_;
	$hash_gene{$key}=$_;
}
close GENE;

# == find gene in var.snpeff
open (SNPEFF,'<',$SNPEFF);
open (OUT,'>',$OUTPUT_GENE);
while (<SNPEFF>) {
	chomp;
	my $line=$_;
	if (/#/) {
		print OUT "$line\n";
		next;
	}
	my @line=split(/\t/,$line);
	my @info=split(/;/,@line[7]);
	my @ANN=split(/,/,@info[$#info]);
	foreach (@ANN) {
		my $ANN=$_;
		my @gene_name=split(/\|/,$ANN);
		my $gene_name=@gene_name[3];
		if (exists $hash_gene{$gene_name}) {
			print OUT "$line\n";
			last;
		}
	}
}
close SNPEFF;
print ">>finish finding the gene in snpeff file\n";
print ">>the output file:$OUTPUT_GENE\n";


# == get the union from var.annodb.indel, var.annodb.snp, var.annovar.clinvar, var.annovar.cosmic70, var.oncotator
# == build index of var.annodb.indel
open (INDEL,'<',$INDEL_ANNODB);
our %hash_indel=();
while (<INDEL>) {
	chomp;
	my $line=$_;
	next if /^Func/;
	$line=~/,chr(\w),(\d+)/;
	my $key_indel=join ("\t",$1,$2);
	$hash_indel{$key_indel}=$line;
}
close INDEL;

# == build index of var.snp.indel
open (SNP,'<',$SNP_ANNODB);
our %hash_snp=();
while (<SNP>) {
	chomp;
	my $line=$_;
        next if /^Func/;
        $line=~/,chr(\w),(\d+)/;
	my $key_snp=join ("\t",$1,$2);
        $hash_snp{$key_snp}=$line;
 }
close SNP;

# == build index of var.clinvar
open (CLINVAR,'<',$CLINVAR_ANNOVAR);
our %hash_clinvar=();
while (<CLINVAR>) {
	chomp;
	my $line=$_;
	my @line=split(/\t/,$line);
	my $key_clinvar=join("\t",@line[10],@line[11]);
	$hash_clinvar{$key_clinvar}=$line;
}
close CLINVAR;

# == build index of var.cosmic70
open(COSMIC70,'<',$COSMIC70_ANNOVAR);
our %hash_cosmic70=();
while (<COSMIC70>) {
	chomp;
	my $line=$_;
	my @line=split(/\t/,$line);
	my $key_cosmic70=join("\t",@line[10],@line[11]);
	$hash_cosmic70{$key_cosmic70}=$line;
}
close COSMIC70;

# == build index of oncotator
open(ONCOTATOR,'<',$ONCOTATOR);
our %hash_oncotator=();
while (<ONCOTATOR>) {
	chomp;
	next if /^#/;
	my $line=$_;
	my @line=split(/\t/,$line);
	my $key_oncotator=join("\t",@line[0],@line[1]);
	$hash_oncotator{$key_oncotator}=$line;
}
close ONCOTATOR;

# == merge all data
open (IN,'<',$OUTPUT_GENE);
open (UNION,'>',$UNION);
while (<IN>) {
	chomp;
	next if /^#/;
	my $line=$_;
	print UNION "snpeff.vcf:\t$line\n";
	my @line=split(/\t/,$line);
	my $cp=join("\t",@line[0],@line[1]);
	if (exists $hash_indel{$cp}) {
		our $indel=$hash_indel{$cp};
	} else {
		our $indel="Not found!";
	}
	print UNION "annodb.indel:\t$indel\n";
	if (exists $hash_snp{$cp}) {
		our $snp=$hash_snp{$cp};
	} else {
		our $snp="Not found!";
	}
	print UNION "annodb.snp:\t$snp\n";
	if (exists $hash_clinvar{$cp}) {
		our $clinvar=$hash_clinvar{$cp};
	} else {
		our $clinvar="Not found!";
	}
	print UNION "anv.clinvar:\t$clinvar\n";
	if (exists $hash_cosmic70{$cp}) {
		our $cosmic70=$hash_cosmic70{$cp};
	} else {
		our $cosmic70="Not found!";
	}
	print UNION "anv.cosmic:\t$cosmic70\n";
	if (exists $hash_oncotator{$cp}) {
		our $oncotator=$hash_oncotator{$cp};
	} else {
		our $oncotator="Not found!";
	}
	print UNION "oncotator:\t$oncotator\n";
	print UNION "\n";
}

# == END
print ">>finish merging all files\n";
print ">>the output file:$UNION\n";
my $final_time=localtime();
print ">>finished at $final_time\n";


















	
	

File: merge/merge.R
# merge gene data
# @wxian 2016Feb22

# == path
GENE="/lustre/project/og04/pub/database/onco_panel/gene"
ANNO="/lustre/project/og04/shenzhongji/human_tumor/QuanlityAssessment/anno/"
OUTPUT="/lustre/project/og04/wangxian/"

# == file_prefix
FILE_1="ck."
FILE_2="ck-VS-mut1."
FILE_3="ck-VS-mut2."
FILE_4="ck-VS-mut3."
FILE_5="ck-VS-mut4."
FILE_6="ck-VS-mut5."
FILE_7="ck-VS-mut6."
FILE_8="ck-VS-mut7."
FILE_9="ck-VS-mut8."
FILE_10="germ1."
FILE_11="germ2."
FILE_12="mut1."
FILE_13="mut2."
FILE_14="mut3."
FILE_15="mut4."
FILE_16="mut5."
FILE_17="mut6."
FILE_18="mut7."
FILE_19="mut8."
FILE=c(FILE_1,FILE_2,FILE_3,FILE_4,FILE_5,FILE_6,FILE_7,FILE_8,FILE_9,FILE_10,FILE_11,FILE_12,FILE_13,FILE_14,FILE_15,FILE_16,FILE_17,FILE_18,FILE_19)
#FILE=c("mut8.")

# == file_format
SV_SNPEFF="sv.snpeff.vcf"
INDEL_ANNODB="var.annodb.indel.exome_summary.csv"
SNP_ANNODB="var.annodb.snp.exome_summary.csv"
CLINVAR_ANNOVAR="var.annovar.hg19_clinvar_20150629_dropped"
COSMIC70_ANNOVAR="var.annovar.hg19_cosmic70_dropped"
VAR_ONCOTATOR="var.oncotator.vcf"
VAR_SNPEFF="var.snpeff.vcf"

gene=data.frame(read.table(GENE))
num.gene=as.numeric(nrow(gene))

for (i in FILE) {
  ## == find gene in var.snpeff.vcf
  file.var.snpeff <- paste0(ANNO,i,VAR_SNPEFF)
  output.name=paste0(OUTPUT,i,VAR_SNPEFF,".gene.csv")
  output=data.frame()
  con <- file(file.var.snpeff,"r")
  line.var.snpeff <- readLines(con,n = 1)
  while( length(line.var.snpeff) != 0 ) {
    if (substr(as.character(line.var.snpeff),1,1) != "#") {
      row.var.snpeff <- strsplit(as.character(line.var.snpeff),"\t")[[1]]
      info.var.snpeff <- strsplit(row.var.snpeff[8],";")[[1]]
      ANN.var.snpeff <- strsplit(info.var.snpeff[length(info.var.snpeff)],",")[[1]]
      for (j in 1:length(ANN.var.snpeff)) {
        gene.var.snpeff <- strsplit(ANN.var.snpeff[j],"\\|")[[1]][4]
        for (k in 1:num.gene) {
          gene.ob=as.character(gene[k,1])
          if (gene.var.snpeff == gene.ob) {
            row=data.frame(line.var.snpeff)
            output=rbind(output,row)
            break
          }
        }
        if (gene.var.snpeff == gene.ob) 
          break 
      }
    }
    line.var.snpeff <- readLines(con, n = 1)
  }
  close(con)
  write.table(output,output.name,row.names=FALSE,col.name=FALSE)
  result=data.frame()
  con1 <- file(output.name,"r")
  line.var.snpeff <- readLines(con1,n = 1)
  while (length(line.var.snpeff) != 0) {
    row.var.snpeff  <- strsplit(as.character(line.var.snpeff ),"\t")[[1]]
    po.ref <- strsplit(row.var.snpeff[2],";")[[1]]
    po.output=data.frame(line.var.snpeff)
    
    # == find position in var.annodb.indel.exome_summary.csv
    file.var.indel.annodb <- paste0(ANNO,i,INDEL_ANNODB)
    con2 <- file(file.var.indel.annodb,"r")
    line.var.indel.annodb <- readLines(con2,n = 1)
    while( length(line.var.indel.annodb) != 0 ) {
      if (substr(as.character(line.var.indel.annodb),1,4) != "Func") {
        row.var.indel.annodb <- strsplit(as.character(line.var.indel.annodb),",")[[1]]
        rank.po=as.numeric(grep("chr1",row.var.indel.annodb)[1])+1
        po.var.indel.annodb <- row.var.indel.annodb[rank.po]
        if (po.var.indel.annodb == po.ref) {
          var.indel.annodb=data.frame(line.var.indel.annodb)
          break
        } else {
          var.indel.annodb=data.frame("line.var.indel.annodb"=c("."))
        }
      }
      line.var.indel.annodb <- readLines(con2, n = 1)
    } 
    po.output=cbind(po.output,var.indel.annodb)
    close(con2)
    
    # == find position in var.annodb.snp.exome_summary.csv
    file.var.snp.annodb <- paste0(ANNO,i,SNP_ANNODB)
    con3 <- file(file.var.snp.annodb,"r")
    line.var.snp.annodb <- readLines(con3,n = 1)
    while( length(line.var.snp.annodb) != 0 ) {
      if (substr(as.character(line.var.snp.annodb),1,4) != "Func") {
        row.var.snp.annodb <- strsplit(as.character(line.var.snp.annodb),",")[[1]]
        rank.po=as.numeric(grep("chr1",row.var.snp.annodb)[1])+1
        po.var.snp.annodb <- row.var.snp.annodb[rank.po]
        if (po.var.snp.annodb == po.ref) {
          var.snp.annodb=data.frame(line.var.snp.annodb)
          break
        } else {
          var.snp.annodb=data.frame("line.var.snp.annodb"=c("."))
        }
      }
      line.var.snp.annodb <- readLines(con3, n = 1)
    }
    po.output=cbind(po.output,var.snp.annodb)
    close(con3)
    
    # == find position in var.annovar.hg19_clinvar_20150629_dropped
    file.var.clinvar.annovar <- paste0(ANNO,i,CLINVAR_ANNOVAR)
    con4 <- file(file.var.clinvar.annovar,"r")
    line.var.clinvar.annovar <- readLines(con4,n = 1)
    while( length(line.var.clinvar.annovar) != 0 ) {
      row.var.clinvar.annovar <- strsplit(as.character(line.var.clinvar.annovar),"\t")[[1]]
      po.var.clinvar.annovar <- row.var.clinvar.annovar[12]
      if (po.var.clinvar.annovar == po.ref) {
        var.clinvar.annovar=data.frame(line.var.clinvar.annovar)
        break
      } else {
        var.clinvar.annovar=data.frame("line.var.clinvar.annovar"=c("."))
      }
      line.var.clinvar.annovar <- readLines(con4, n = 1)
    }
    po.output=cbind(po.output,var.clinvar.annovar)
    close(con4)
    
    # == find position in var.annovar.hg19_cosmic70_dropped
    file.var.cosmic70.annovar <- paste0(ANNO,i,COSMIC70_ANNOVAR)
    con5 <- file(file.var.cosmic70.annovar,"r")
    line.var.cosmic70.annovar <- readLines(con5,n = 1)
    while( length(line.var.cosmic70.annovar) != 0 ) {
      row.var.cosmic70.annovar <- strsplit(as.character(line.var.cosmic70.annovar),"\t")[[1]]
      po.var.cosmic70.annovar <- row.var.cosmic70.annovar[12]
      if (po.var.cosmic70.annovar == po.ref) {
        var.cosmic70.annovar=data.frame(line.var.cosmic70.annovar)
        break
      } else {
        var.cosmic70.annovar=data.frame("line.var.cosmic70.annovar"=c("."))
      }
      line.var.cosmic70.annovar <- readLines(con5, n = 1)
    }
    po.output=cbind(po.output,var.cosmic70.annovar)
    close(con5)
    
    # == find position in var.oncotator.vcf
    file.var.oncotator.vcf <- paste0(ANNO,i,VAR_ONCOTATOR)
    con6 <- file(file.var.oncotator.vcf,"r")
    line.var.oncotator.vcf <- readLines(con6,n = 1)
    while( length(line.var.oncotator.vcf) != 0 ) {
      if (substr(as.character(line.var.oncotator.vcf),1,1) != "#") {
        row.var.oncotator.vcf <- strsplit(as.character(line.var.oncotator.vcf),"\t")[[1]]
        po.var.oncotator.vcf <- row.var.oncotator.vcf[2]
        if (po.var.oncotator.vcf == po.ref) {
          var.oncotator.vcf=data.frame(line.var.oncotator.vcf)
          break
        } else {
          var.oncotator.vcf=data.frame("line.var.oncotator.vcf"=c("."))
        } 
      }
      line.var.oncotator.vcf <- readLines(con6, n = 1)
    }
    po.output=cbind(po.output,var.oncotator.vcf)
    close(con6)  
    result=rbind(result,po.output)
    line.var.snpeff <- readLines(con1,n = 1)
  }
  close(con1)
}
result.name=paste0(OUTPUT,i,"merge.csv")
write.csv(result,result.name,row.names=FALSE)